runtime/drivers/olap.go

```diff
@@ -495,6 +500,128 @@ func (d Dialect) DateDiff(grain runtimev1.TimeGrain, t1, t2 time.Time) (string,
+				prefix = "SELECT * FROM (VALUES "
+				suffix = "t("
+			}
+			if rows > 0 {
+				prefix += ", "
+			}
+		} else {
+			// format - select 1 as a, 2 as b union all select 3 as a, 4 as b
+			if rows > 0 {
+				prefix += " UNION ALL "
+			}
+			prefix += "SELECT "
+		}
+
+		dimVals = append(dimVals, values[0])
+		for i, v := range values {
+			if d == DialectDruid {
+				if i == 0 {
+					prefix += "("
+				}
+				if rows == 0 {
+					suffix += d.EscapeIdentifier(result.Schema.Fields[i].Name)
+					if i != len(result.Schema.Fields)-1 {
+						suffix += ", "
+					}
+				}
+			}
+			if i > 0 {
+				prefix += ", "
+			}
+			ok, expr, err := d.GetValExpr(v, result.Schema.Fields[i].Type.Code)
+			if err != nil {
+				return "", nil, fmt.Errorf("select inline: failed to get value expression: %w", err)
+			}
+			if !ok {
+				return "", nil, fmt.Errorf("select inline: unsupported value type %q: %w", result.Schema.Fields[i].Type.Code, ErrOptimizationFailure)
+			}
+			if d == DialectDruid {
+				prefix += expr
+			} else {
+				prefix += fmt.Sprintf("%s AS %s", expr, d.EscapeIdentifier(result.Schema.Fields[i].Name))
+			}
+		}
+
+		if d == DialectDruid {
+			prefix += ")"
+			if rows == 0 {
+				suffix += ")"
+			}
+		}
+
+		rows++
+	}
+
+	if d == DialectDruid {
+		prefix += ") "
+	}
+
+	return prefix + suffix, dimVals, nil
+}
+
+func (d Dialect) GetValExpr(val any, typ runtimev1.Type_Code) (bool, string, error) {
+	if val == nil {
+		return true, "NULL", nil
+	}
+	switch typ {
+	case runtimev1.Type_CODE_STRING:
+		return true, fmt.Sprintf("'%v'", val), nil
+	case runtimev1.Type_CODE_INT8, runtimev1.Type_CODE_INT16, runtimev1.Type_CODE_INT32, runtimev1.Type_CODE_INT64, runtimev1.Type_CODE_INT128, runtimev1.Type_CODE_INT256, runtimev1.Type_CODE_UINT8, runtimev1.Type_CODE_UINT16, runtimev1.Type_CODE_UINT32, runtimev1.Type_CODE_UINT64, runtimev1.Type_CODE_UINT128, runtimev1.Type_CODE_UINT256, runtimev1.Type_CODE_FLOAT32, runtimev1.Type_CODE_FLOAT64, runtimev1.Type_CODE_DECIMAL:
+		// check NaN and Inf
+		if f, ok := val.(float64); ok && (math.IsNaN(f) || math.IsInf(f, 0)) {
+			return true, "NULL", nil
+		}
+
+		return true, fmt.Sprintf("%v", val), nil
+	case runtimev1.Type_CODE_BOOL:
+		return true, fmt.Sprintf("%v", val), nil
+	case runtimev1.Type_CODE_TIME, runtimev1.Type_CODE_DATE, runtimev1.Type_CODE_TIMESTAMP:
+		if t, ok := val.(time.Time); ok {
+			if ok, expr := d.GetTimeExpr(t); ok {
+				return true, expr, nil
+			}
+			return false, "", fmt.Errorf("cannot get time expr for dialect %q", d)
+		}
+		return false, "", fmt.Errorf("unsupported time type %q", typ)
+	default:
+		return false, "", fmt.Errorf("unsupported type %q", typ)
+	}
+}
+
+func (d Dialect) GetTimeExpr(t time.Time) (bool, string) {
+	switch d {
+	case DialectClickHouse:
+		return true, fmt.Sprintf("parseDateTimeBestEffort('%s')", t.Format(time.RFC3339))
+	case DialectDuckDB, DialectDruid, DialectPinot:
+		return true, fmt.Sprintf("CAST('%s' AS TIMESTAMP)", t.Format(time.RFC3339))
```

Comment:
Should it use `time.RFC3339Nano`? The databases usually have at least millisecond precision, but `RFC3339` only has second precision IIRC